Deduplicate the world state implementations

sammy-tri opened the pull request:
Moves some code out of dev/ which will go into the eventual pick-and-place CI test.

Note to platform reviewer:  The code in dev/ can probably be skipped entirely for platform review (though go ahead if you want, I guess).  The code in kuka_iiwa_arm/pick_and_place_test should be reviewed as-new, since it hasn't yet had proper platform review.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/robotlocomotion/drake/6199)
<!-- Reviewable:end -->

sammy-tri commented:
+@siyuanfeng-tri for platform review

Cc: @naveenoid (I know you're busy this week so no obligation to review, but as the original author of some of this code it may be somewhat relevant to your interests...   or not.  :)

---

Review status: 0 of 14 files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6199#-:-KlOoLGSDlBqLFhXcMLe:b-mb6kx3)*
<!-- Sent from Reviewable.io -->

sammy-tri commented:
@drake-jenkins-bot linux-gcc-bazel-experimental please
@drake-jenkins-bot linux-xenial-clang-experimental-minimal please

---

Review status: 0 of 14 files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6199#-:-KlOu96F_ZS14JKdIJlZ:b-1pv530)*
<!-- Sent from Reviewable.io -->

siyuanfeng-tri commented:
thanks sammy. 
high level question, why is the folder called pick_and_place_test? rather than something like pick and place? 

---

Reviewed 16 of 16 files at r1.
Review status: all files reviewed at latest revision, 9 unresolved discussions.

---

*[drake/examples/kuka_iiwa_arm/dev/pick_and_place/pick_and_place_state_machine.cc, line 36 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6199#-KlOsvQXGv3t53OaLgIh:-KlOsvQXGv3t53OaLgIi:bysy35g) ([raw file](https://github.com/robotlocomotion/drake/blob/17b1472babd3be5915c4d99f30563ef3d103da05/drake/examples/kuka_iiwa_arm/dev/pick_and_place/pick_and_place_state_machine.cc#L36)):*
> ```cpp
>   DRAKE_NO_COPY_NO_MOVE_NO_ASSIGN(WorldStateSubscriber)
> 
>   WorldStateSubscriber(lcm::LCM* lcm, WorldState* state)
> ```

why doesn't this own the world state? 

---

*[drake/examples/kuka_iiwa_arm/pick_and_place_test/CMakeLists.txt, line 7 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6199#-KlOsRKnE5uiqxbnpHoy:-KlOsRKnE5uiqxbnpHoz:b-mchm1z) ([raw file](https://github.com/robotlocomotion/drake/blob/17b1472babd3be5915c4d99f30563ef3d103da05/drake/examples/kuka_iiwa_arm/pick_and_place_test/CMakeLists.txt#L7)):*
> ```CMake
>       drakeCommon
>       drakeKukaIiwaArmCommon
>       drakeLcmSystem
> ```

do we need this?

---

*[drake/examples/kuka_iiwa_arm/pick_and_place_test/world_state.cc, line 34 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6199#-KlOrYm3BjFPhT32do32:-KlOrYm3BjFPhT32do33:b-iqnlk3) ([raw file](https://github.com/robotlocomotion/drake/blob/17b1472babd3be5915c4d99f30563ef3d103da05/drake/examples/kuka_iiwa_arm/pick_and_place_test/world_state.cc#L34)):*
> ```cpp
> WorldState::~WorldState() { }
> 
> void WorldState::HandleIiwaStatus(const bot_core::robot_state_t* iiwa_msg) {
> ```

why not &? 

---

*[drake/examples/kuka_iiwa_arm/pick_and_place_test/world_state.cc, line 62 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6199#-KlOrd8Rf_8OnI49efj7:-KlOrd8Rf_8OnI49efj8:b3kqp) ([raw file](https://github.com/robotlocomotion/drake/blob/17b1472babd3be5915c4d99f30563ef3d103da05/drake/examples/kuka_iiwa_arm/pick_and_place_test/world_state.cc#L62)):*
> ```cpp
> }
> 
> void WorldState::HandleWsgStatus(const lcmt_schunk_wsg_status* wsg_msg) {
> ```

&

---

*[drake/examples/kuka_iiwa_arm/pick_and_place_test/world_state.cc, line 84 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6199#-KlOre05O4KUzf-4FShz:-KlOre05O4KUzf-4FSi-:b3kqp) ([raw file](https://github.com/robotlocomotion/drake/blob/17b1472babd3be5915c4d99f30563ef3d103da05/drake/examples/kuka_iiwa_arm/pick_and_place_test/world_state.cc#L84)):*
> ```cpp
> }
> 
> void WorldState::HandleObjectStatus(const bot_core::robot_state_t* obj_msg) {
> ```

&

---

*[drake/examples/kuka_iiwa_arm/pick_and_place_test/world_state.h, line 29 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6199#-KlOvIoWwgNlXTYwaVSA:-KlOvIoWwgNlXTYwaVSB:bgcnh4c) ([raw file](https://github.com/robotlocomotion/drake/blob/17b1472babd3be5915c4d99f30563ef3d103da05/drake/examples/kuka_iiwa_arm/pick_and_place_test/world_state.h#L29)):*
> ```objc
>   /**
>    * Constructs an WorldState object that holds the states that represent a pick
>    * and place scenario.
> ```

should add a sentence about how the internal rbt is initialized, also end_effector_name is link name. 

---

*[drake/examples/kuka_iiwa_arm/pick_and_place_test/world_state.h, line 70 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6199#-KlOuIKpUGRGZkJ8fFad:-KlOuIKpUGRGZkJ8fFae:b2ivzck) ([raw file](https://github.com/robotlocomotion/drake/blob/17b1472babd3be5915c4d99f30563ef3d103da05/drake/examples/kuka_iiwa_arm/pick_and_place_test/world_state.h#L70)):*
> ```objc
>   // located. Since this information comes from LCM and thus may be delayed,
>   // it's easier for us to own a model internally.
>   std::unique_ptr<RigidBodyTree<double>> iiwa_;
> ```

I am debating whether this class should have 2 flavors, one is exactly like this: we initialize the rbt's base using the lcm message's pose. We should at least copy this comment to the constructor so it's more obvious. 

The other would be just use a RBT& passed in. Because the first flavor is pretty subtle, if some body already setup the RBT, but didn't make the message correctly (containing the base offset), his kinematics would be all wrong. 

---

*[drake/examples/kuka_iiwa_arm/pick_and_place_test/world_state.h, line 72 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6199#-KlOv5mUtO_w8lshI99_:-KlOv5mUtO_w8lshI99a:bco2n7e) ([raw file](https://github.com/robotlocomotion/drake/blob/17b1472babd3be5915c4d99f30563ef3d103da05/drake/examples/kuka_iiwa_arm/pick_and_place_test/world_state.h#L72)):*
> ```objc
>   std::unique_ptr<RigidBodyTree<double>> iiwa_;
>   const std::string iiwa_model_path_;
>   const std::string ee_name_;
> ```

If I were a real platform reviewer, I would probably complain about the ee part. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6199#-:-KlOtoX4cAhzWGdoq9Sg:btt47u4)*
<!-- Sent from Reviewable.io -->

siyuanfeng-tri commented:
Review status: all files reviewed at latest revision, 11 unresolved discussions.

---

*[drake/examples/kuka_iiwa_arm/pick_and_place_test/world_state.h, line 29 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6199#-KlOwyZEfbm1yvxaMs84:-KlOwyZEfbm1yvxaMs85:b2noa9m) ([raw file](https://github.com/robotlocomotion/drake/blob/17b1472babd3be5915c4d99f30563ef3d103da05/drake/examples/kuka_iiwa_arm/pick_and_place_test/world_state.h#L29)):*
> ```objc
>   /**
>    * Constructs an WorldState object that holds the states that represent a pick
>    * and place scenario.
> ```

I would also add a comment saying these states are not synchronized in anyway. Simply the latest status. 

---

*[drake/examples/kuka_iiwa_arm/pick_and_place_test/world_state.h, line 34 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6199#-KlOxM5tsEaWjIGhytRr:-KlOxM5tsEaWjIGhytRs:b-6gemh2) ([raw file](https://github.com/robotlocomotion/drake/blob/17b1472babd3be5915c4d99f30563ef3d103da05/drake/examples/kuka_iiwa_arm/pick_and_place_test/world_state.h#L34)):*
> ```objc
>              const std::string& end_effector_name);
> 
>   virtual ~WorldState();
> ```

probably dont need this yet? 

---

*[drake/examples/kuka_iiwa_arm/pick_and_place_test/world_state.h, line 37 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6199#-KlOxH-gwZDEMm0ivqXn:-KlOxH-gwZDEMm0ivqXo:bm1ca26) ([raw file](https://github.com/robotlocomotion/drake/blob/17b1472babd3be5915c4d99f30563ef3d103da05/drake/examples/kuka_iiwa_arm/pick_and_place_test/world_state.h#L37)):*
> ```objc
> 
>   // Handles iiwa states from the LCM message.
>   void HandleIiwaStatus(const bot_core::robot_state_t* iiwa_msg);
> ```

These public methods probably need a one line doxygen comment.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6199)*
<!-- Sent from Reviewable.io -->

sammy-tri commented:
I think because my first thought was that it was explicitly to build the CI test out of.  Renamed (though it's gonna wreak havoc on the rebase for my in-flight patches, removing _test did seem better)

---

Review status: 3 of 14 files reviewed at latest revision, 11 unresolved discussions.

---

*[drake/examples/kuka_iiwa_arm/dev/pick_and_place/pick_and_place_state_machine.cc, line 36 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6199#-KlOsvQXGv3t53OaLgIh:-KlOwkXme5CQ8iBGTfvK:b-im15h6) ([raw file](https://github.com/robotlocomotion/drake/blob/17b1472babd3be5915c4d99f30563ef3d103da05/drake/examples/kuka_iiwa_arm/dev/pick_and_place/pick_and_place_state_machine.cc#L36)):*
<details><summary><i>Previously, siyuanfeng-tri wrote…</i></summary><blockquote>

why doesn't this own the world state? 
</blockquote></details>

I think it would be fine for this to own the world state and have a const ref accessor...   then env_state in RunPickAndPlaceDemo could be a ref back into this class.   I don't think I really prefer one to the other (also the monolithic / system based demo uses WorldState without a subscriber, which makes no difference here but might have been part of my reasoning).  

I suppose also iiwa_end_effector_name is defined locally in RunPickAndPlaceDemo and passed into the WorldState constructor, but I could make that a kConstant.

In retrospect, I probably also could have accomplished my goal with lambdas instead of creating a helper class, but I'm not sure that failing to do so is a defect.

---

*[drake/examples/kuka_iiwa_arm/pick_and_place_test/CMakeLists.txt, line 7 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6199#-KlOsRKnE5uiqxbnpHoy:-KlOxaCmmXeW4aKV_4Su:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/17b1472babd3be5915c4d99f30563ef3d103da05/drake/examples/kuka_iiwa_arm/pick_and_place_test/CMakeLists.txt#L7)):*
<details><summary><i>Previously, siyuanfeng-tri wrote…</i></summary><blockquote>

do we need this?
</blockquote></details>

Done.

---

*[drake/examples/kuka_iiwa_arm/pick_and_place_test/world_state.cc, line 34 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6199#-KlOrYm3BjFPhT32do32:-KlOxbwjNEnvWxbqhzT6:b-jb2w82) ([raw file](https://github.com/robotlocomotion/drake/blob/17b1472babd3be5915c4d99f30563ef3d103da05/drake/examples/kuka_iiwa_arm/pick_and_place_test/world_state.cc#L34)):*
<details><summary><i>Previously, siyuanfeng-tri wrote…</i></summary><blockquote>

why not &? 
</blockquote></details>

It's a pointer in the callback from LCM which we then invoke here....   which is probably why it was always a pointer in the call to this function before I removed the first two arguments.

---

*[drake/examples/kuka_iiwa_arm/pick_and_place_test/world_state.cc, line 62 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6199#-KlOrd8Rf_8OnI49efj7:-KlOy67Ws78pELXWwe8e:b-34vmyd) ([raw file](https://github.com/robotlocomotion/drake/blob/17b1472babd3be5915c4d99f30563ef3d103da05/drake/examples/kuka_iiwa_arm/pick_and_place_test/world_state.cc#L62)):*
<details><summary><i>Previously, siyuanfeng-tri wrote…</i></summary><blockquote>

&
</blockquote></details>

See above

---

*[drake/examples/kuka_iiwa_arm/pick_and_place_test/world_state.cc, line 84 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6199#-KlOre05O4KUzf-4FShz:-KlOy7LGgA0M7gVN86Q3:b-34vmyd) ([raw file](https://github.com/robotlocomotion/drake/blob/17b1472babd3be5915c4d99f30563ef3d103da05/drake/examples/kuka_iiwa_arm/pick_and_place_test/world_state.cc#L84)):*
<details><summary><i>Previously, siyuanfeng-tri wrote…</i></summary><blockquote>

&
</blockquote></details>

See above

---

*[drake/examples/kuka_iiwa_arm/pick_and_place_test/world_state.h, line 29 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6199#-KlOvIoWwgNlXTYwaVSA:-KlOyNDQ8raDbQqvuq4d:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/17b1472babd3be5915c4d99f30563ef3d103da05/drake/examples/kuka_iiwa_arm/pick_and_place_test/world_state.h#L29)):*
<details><summary><i>Previously, siyuanfeng-tri wrote…</i></summary><blockquote>

should add a sentence about how the internal rbt is initialized, also end_effector_name is link name. 
</blockquote></details>

Done.

---

*[drake/examples/kuka_iiwa_arm/pick_and_place_test/world_state.h, line 29 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6199#-KlOwyZEfbm1yvxaMs84:-KlOyWyGgKH2DNuu_YeB:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/17b1472babd3be5915c4d99f30563ef3d103da05/drake/examples/kuka_iiwa_arm/pick_and_place_test/world_state.h#L29)):*
<details><summary><i>Previously, siyuanfeng-tri wrote…</i></summary><blockquote>

I would also add a comment saying these states are not synchronized in anyway. Simply the latest status. 
</blockquote></details>

Done.

---

*[drake/examples/kuka_iiwa_arm/pick_and_place_test/world_state.h, line 34 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6199#-KlOxM5tsEaWjIGhytRr:-KlOybtThQ3-Fr0VqXC9:b27abc5) ([raw file](https://github.com/robotlocomotion/drake/blob/17b1472babd3be5915c4d99f30563ef3d103da05/drake/examples/kuka_iiwa_arm/pick_and_place_test/world_state.h#L34)):*
<details><summary><i>Previously, siyuanfeng-tri wrote…</i></summary><blockquote>

probably dont need this yet? 
</blockquote></details>

It certainly didn't need to be virtual.  I don't think there's any harm in defining the destructor explicitly so that there's only one version in world_state.o instead of being generated inline wherever the class is instantiated.

---

*[drake/examples/kuka_iiwa_arm/pick_and_place_test/world_state.h, line 37 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6199#-KlOxH-gwZDEMm0ivqXn:-KlOz0_t_4ajI-ScUhWw:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/17b1472babd3be5915c4d99f30563ef3d103da05/drake/examples/kuka_iiwa_arm/pick_and_place_test/world_state.h#L37)):*
<details><summary><i>Previously, siyuanfeng-tri wrote…</i></summary><blockquote>

These public methods probably need a one line doxygen comment.
</blockquote></details>

Done.

---

*[drake/examples/kuka_iiwa_arm/pick_and_place_test/world_state.h, line 70 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6199#-KlOuIKpUGRGZkJ8fFad:-KlP0YeEsjY37a5ISLI1:b723ynv) ([raw file](https://github.com/robotlocomotion/drake/blob/17b1472babd3be5915c4d99f30563ef3d103da05/drake/examples/kuka_iiwa_arm/pick_and_place_test/world_state.h#L70)):*
<details><summary><i>Previously, siyuanfeng-tri wrote…</i></summary><blockquote>

I am debating whether this class should have 2 flavors, one is exactly like this: we initialize the rbt's base using the lcm message's pose. We should at least copy this comment to the constructor so it's more obvious. 

The other would be just use a RBT& passed in. Because the first flavor is pretty subtle, if some body already setup the RBT, but didn't make the message correctly (containing the base offset), his kinematics would be all wrong. 

</blockquote></details>

I thought about doing something like that, actually.  I'd have to figure out how to make get_iiwa_base() work in the case where the RBT is passed in but I suspect that's fairly simple (or I could make the base a ctor argument, which is even easier).

If think if that gets done in this version of the PR it's worth adding a unit test (it might be worth it anyway to check the kinematics calculations)

---

*[drake/examples/kuka_iiwa_arm/pick_and_place_test/world_state.h, line 72 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6199#-KlOv5mUtO_w8lshI99_:-KlP1OF2P6_z_UqgTN_s:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/17b1472babd3be5915c4d99f30563ef3d103da05/drake/examples/kuka_iiwa_arm/pick_and_place_test/world_state.h#L72)):*
<details><summary><i>Previously, siyuanfeng-tri wrote…</i></summary><blockquote>

If I were a real platform reviewer, I would probably complain about the ee part. 
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6199#-:-KlP1PfgDbKlVubOCWQv:bpr7046)*
<!-- Sent from Reviewable.io -->

siyuanfeng-tri commented:
Review status: 3 of 14 files reviewed at latest revision, 6 unresolved discussions.

---

*[drake/examples/kuka_iiwa_arm/dev/pick_and_place/pick_and_place_state_machine.cc, line 36 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6199#-KlOsvQXGv3t53OaLgIh:-KlP9GHiGxvmLHHJA7pT:b-yykab7) ([raw file](https://github.com/robotlocomotion/drake/blob/17b1472babd3be5915c4d99f30563ef3d103da05/drake/examples/kuka_iiwa_arm/dev/pick_and_place/pick_and_place_state_machine.cc#L36)):*
<details><summary><i>Previously, sammy-tri (Sam Creasey) wrote…</i></summary><blockquote>

I think it would be fine for this to own the world state and have a const ref accessor...   then env_state in RunPickAndPlaceDemo could be a ref back into this class.   I don't think I really prefer one to the other (also the monolithic / system based demo uses WorldState without a subscriber, which makes no difference here but might have been part of my reasoning).  

I suppose also iiwa_end_effector_name is defined locally in RunPickAndPlaceDemo and passed into the WorldState constructor, but I could make that a kConstant.

In retrospect, I probably also could have accomplished my goal with lambdas instead of creating a helper class, but I'm not sure that failing to do so is a defect.
</blockquote></details>

Sure, not a big deal, this is in dev anyway. 

---

*[drake/examples/kuka_iiwa_arm/pick_and_place/world_state.cc, line 34 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6199#-KlOrYm3BjFPhT32do32:-KlP81gSM8sx9iT6GbVb:b-9fzpt8) ([raw file](https://github.com/robotlocomotion/drake/blob/17b1472babd3be5915c4d99f30563ef3d103da05/drake/examples/kuka_iiwa_arm/pick_and_place/world_state.cc#L34)):*
<details><summary><i>Previously, sammy-tri (Sam Creasey) wrote…</i></summary><blockquote>

It's a pointer in the callback from LCM which we then invoke here....   which is probably why it was always a pointer in the call to this function before I removed the first two arguments.
</blockquote></details>

yeah i know, but we practically always treat it as a const& anyway. might as well just use const&, and have the lcm callback do the *.

---

*[drake/examples/kuka_iiwa_arm/pick_and_place/world_state.h, line 29 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6199#-KlOvIoWwgNlXTYwaVSA:-KlP90PgMMUzB-hEmYIO:b-s4orvm) ([raw file](https://github.com/robotlocomotion/drake/blob/17b1472babd3be5915c4d99f30563ef3d103da05/drake/examples/kuka_iiwa_arm/pick_and_place/world_state.h#L29)):*
<details><summary><i>Previously, sammy-tri (Sam Creasey) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

I'd add another sentence saying RBT's base will be from the first received lcm message. 

---

*[drake/examples/kuka_iiwa_arm/pick_and_place/world_state.h, line 70 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6199#-KlOuIKpUGRGZkJ8fFad:-KlP8US5FOKY25kzTT3a:b-y7lljq) ([raw file](https://github.com/robotlocomotion/drake/blob/17b1472babd3be5915c4d99f30563ef3d103da05/drake/examples/kuka_iiwa_arm/pick_and_place/world_state.h#L70)):*
<details><summary><i>Previously, sammy-tri (Sam Creasey) wrote…</i></summary><blockquote>

I thought about doing something like that, actually.  I'd have to figure out how to make get_iiwa_base() work in the case where the RBT is passed in but I suspect that's fairly simple (or I could make the base a ctor argument, which is even easier).

If think if that gets done in this version of the PR it's worth adding a unit test (it might be worth it anyway to check the kinematics calculations)
</blockquote></details>

i know how to do the get iiwa base thing for the second case. https://github.com/RobotLocomotion/drake/blob/master/drake/manipulation/util/robot_state_msg_translator.cc#L276
root_joint is the fixed joint between world and the first link. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6199)*
<!-- Sent from Reviewable.io -->

naveenoid commented:
Reviewed 4 of 20 files at r1, 9 of 15 files at r2.
Review status: 11 of 14 files reviewed at latest revision, 6 unresolved discussions.

---

*[drake/examples/kuka_iiwa_arm/pick_and_place/world_state.cc, line 34 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6199#-KlOrYm3BjFPhT32do32:-KlPBaLi4ZhLscSVnnkm:b-3105xt) ([raw file](https://github.com/robotlocomotion/drake/blob/17b1472babd3be5915c4d99f30563ef3d103da05/drake/examples/kuka_iiwa_arm/pick_and_place/world_state.cc#L34)):*
<details><summary><i>Previously, siyuanfeng-tri wrote…</i></summary><blockquote>

yeah i know, but we practically always treat it as a const& anyway. might as well just use const&, and have the lcm callback do the *.
</blockquote></details>

+1 Id prefer to keep the callback'y things and the processing things independent. 

---

*[drake/examples/kuka_iiwa_arm/pick_and_place/world_state.h, line 62 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/6199#-KlPCXOqqELSvS-6T1dm:-KlPCXOqqELSvS-6T1dn:b-4fe7s) ([raw file](https://github.com/robotlocomotion/drake/blob/f7c2d009fbe29ea7238d8f7a928c44286ed12f2e/drake/examples/kuka_iiwa_arm/pick_and_place/world_state.h#L62)):*
> ```objc
>   }
>   const Vector6<double>& get_iiwa_end_effector_velocity() const {
>     return iiwa_end_effector_vel_;
> ```

BTW: might as well call this velocity. <meh>

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6199)*
<!-- Sent from Reviewable.io -->

naveenoid commented:
Reviewed 1 of 20 files at r1, 2 of 15 files at r2.
Review status: 12 of 14 files reviewed at latest revision, 6 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6199)*
<!-- Sent from Reviewable.io -->

siyuanfeng-tri commented:
*[drake/examples/kuka_iiwa_arm/pick_and_place/world_state.h, line 62 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/6199#-KlPCXOqqELSvS-6T1dm:-KlPDCl_icoxDmOQYKRY:b-7wwwlw) ([raw file](https://github.com/robotlocomotion/drake/blob/f7c2d009fbe29ea7238d8f7a928c44286ed12f2e/drake/examples/kuka_iiwa_arm/pick_and_place/world_state.h#L62)):*
<details><summary><i>Previously, naveenoid (Naveen Kuppuswamy) wrote…</i></summary><blockquote>

BTW: might as well call this velocity. <meh>
</blockquote></details>

BTW: neverrrrrrrr

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6199)*
<!-- Sent from Reviewable.io -->

sammy-tri commented:
Review status: 8 of 16 files reviewed at latest revision, 5 unresolved discussions.

---

*[drake/examples/kuka_iiwa_arm/pick_and_place/world_state.cc, line 34 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6199#-KlOrYm3BjFPhT32do32:-KlPKCIvxfE1RJgEP8RB:b-w930b0) ([raw file](https://github.com/robotlocomotion/drake/blob/17b1472babd3be5915c4d99f30563ef3d103da05/drake/examples/kuka_iiwa_arm/pick_and_place/world_state.cc#L34)):*
<details><summary><i>Previously, naveenoid (Naveen Kuppuswamy) wrote…</i></summary><blockquote>

+1 Id prefer to keep the callback'y things and the processing things independent. 
</blockquote></details>

I will admit, with great disappointment, that this use of `const *` does not obviously meet the allowances in https://google.github.io/styleguide/cppguide.html#Reference_Arguments -- I personally think pointers should always be acceptable instead of references (and I *hate* stuffing a pointer into a reference since that's how the horror of null references comes about), but here we are.

Done.

---

*[drake/examples/kuka_iiwa_arm/pick_and_place/world_state.cc, line 62 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6199#-KlOrd8Rf_8OnI49efj7:-KlPZnJNk53ozkQ-vQQD:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/17b1472babd3be5915c4d99f30563ef3d103da05/drake/examples/kuka_iiwa_arm/pick_and_place/world_state.cc#L62)):*
<details><summary><i>Previously, sammy-tri (Sam Creasey) wrote…</i></summary><blockquote>

See above
</blockquote></details>

Done.

---

*[drake/examples/kuka_iiwa_arm/pick_and_place/world_state.cc, line 84 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6199#-KlOre05O4KUzf-4FShz:-KlPZnrbJU5_YPdHSd2L:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/17b1472babd3be5915c4d99f30563ef3d103da05/drake/examples/kuka_iiwa_arm/pick_and_place/world_state.cc#L84)):*
<details><summary><i>Previously, sammy-tri (Sam Creasey) wrote…</i></summary><blockquote>

See above
</blockquote></details>

Done.

---

*[drake/examples/kuka_iiwa_arm/pick_and_place/world_state.h, line 29 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6199#-KlOvIoWwgNlXTYwaVSA:-KlPJf0yIKBy_Ujz2juB:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/17b1472babd3be5915c4d99f30563ef3d103da05/drake/examples/kuka_iiwa_arm/pick_and_place/world_state.h#L29)):*
<details><summary><i>Previously, siyuanfeng-tri wrote…</i></summary><blockquote>

I'd add another sentence saying RBT's base will be from the first received lcm message. 
</blockquote></details>

Done.

---

*[drake/examples/kuka_iiwa_arm/pick_and_place/world_state.h, line 70 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6199#-KlOuIKpUGRGZkJ8fFad:-KlPZc7CUeyFrdwdPAOQ:b-sbqhx6) ([raw file](https://github.com/robotlocomotion/drake/blob/17b1472babd3be5915c4d99f30563ef3d103da05/drake/examples/kuka_iiwa_arm/pick_and_place/world_state.h#L70)):*
<details><summary><i>Previously, siyuanfeng-tri wrote…</i></summary><blockquote>

i know how to do the get iiwa base thing for the second case. https://github.com/RobotLocomotion/drake/blob/master/drake/manipulation/util/robot_state_msg_translator.cc#L276
root_joint is the fixed joint between world and the first link. 
</blockquote></details>

Added a unit test.  Are we OK with a TODO for the alternate constructor form?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6199)*
<!-- Sent from Reviewable.io -->

siyuanfeng-tri commented:
Review status: 8 of 16 files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[drake/examples/kuka_iiwa_arm/pick_and_place/world_state.h, line 70 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6199#-KlOuIKpUGRGZkJ8fFad:-KlP_PbQbbdAJLRjxpNw:b-8zb3ww) ([raw file](https://github.com/robotlocomotion/drake/blob/17b1472babd3be5915c4d99f30563ef3d103da05/drake/examples/kuka_iiwa_arm/pick_and_place/world_state.h#L70)):*
<details><summary><i>Previously, sammy-tri (Sam Creasey) wrote…</i></summary><blockquote>

Added a unit test.  Are we OK with a TODO for the alternate constructor form?
</blockquote></details>

sure. although I am not seeing the todo, maybe you haven't added it? 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6199)*
<!-- Sent from Reviewable.io -->

siyuanfeng-tri commented:
Reviewed 2 of 20 files at r1, 2 of 15 files at r2, 2 of 8 files at r3.
Review status: 10 of 16 files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6199)*
<!-- Sent from Reviewable.io -->

sammy-tri commented:
Review status: 9 of 16 files reviewed at latest revision, 1 unresolved discussion.

---

*[drake/examples/kuka_iiwa_arm/pick_and_place/world_state.h, line 70 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6199#-KlOuIKpUGRGZkJ8fFad:-KlPaR6nDLTwIb5grUt_:b-jhm6rt) ([raw file](https://github.com/robotlocomotion/drake/blob/17b1472babd3be5915c4d99f30563ef3d103da05/drake/examples/kuka_iiwa_arm/pick_and_place/world_state.h#L70)):*
<details><summary><i>Previously, siyuanfeng-tri wrote…</i></summary><blockquote>

sure. although I am not seeing the todo, maybe you haven't added it? 
</blockquote></details>

TODO Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6199)*
<!-- Sent from Reviewable.io -->

sammy-tri commented:
+@ggould-tri as tomorrow's on-call platform reviewer

---

Review status: 9 of 16 files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6199#-:-KlPbO7pc_sxZ56PLg5s:b91ci37)*
<!-- Sent from Reviewable.io -->

ggould-tri commented:
With as-new'ing the stuff moved out of dev, this is a very long PR.  It is likely to take me a while to complete platform review.

---

Reviewed 2 of 20 files at r1, 6 of 15 files at r2, 1 of 8 files at r3.
Review status: 10 of 16 files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6199#-:-KlPuP_0PhzNtkYteAKa:b-7ksegk)*
<!-- Sent from Reviewable.io -->

ggould-tri commented:
Ah, never mind, the really big one is still in dev.  Should be able to finish this tomorrow morning.

---

Review status: 10 of 16 files reviewed at latest revision, 2 unresolved discussions.

---

*[drake/examples/kuka_iiwa_arm/dev/pick_and_place/pick_and_place_state_machine.cc, line 37 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/6199#-KlPwX7qu00RnKavMQHt:-KlPwX8G3-KOWiJnlwhk:bvkl92p) ([raw file](https://github.com/robotlocomotion/drake/blob/05d27e3933ebc07da0bdeeec80b3d3f61ef67ee1/drake/examples/kuka_iiwa_arm/dev/pick_and_place/pick_and_place_state_machine.cc#L37)):*
> ```cpp
> 
>   WorldStateSubscriber(lcm::LCM* lcm, WorldState* state)
>   : lcm_(lcm),
> ```

FYI -- This indentation is not GSG standard.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6199#-:-KlPwqU_2F2zNoo598LB:b-g5r1jf)*
<!-- Sent from Reviewable.io -->

ggould-tri commented:
Reviewed 2 of 20 files at r1, 4 of 15 files at r2, 5 of 8 files at r3, 2 of 2 files at r4.
Review status: all files reviewed at latest revision, 7 unresolved discussions.

---

*[drake/examples/kuka_iiwa_arm/pick_and_place/BUILD, line 8 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/6199#-KlPyWr0bw1MNRdfgm2x:-KlPyWr0bw1MNRdfgm2y:b-hmbim5) ([raw file](https://github.com/robotlocomotion/drake/blob/05d27e3933ebc07da0bdeeec80b3d3f61ef67ee1/drake/examples/kuka_iiwa_arm/pick_and_place/BUILD#L8)):*
> ```Python
>     "//tools:drake.bzl",
>     "drake_cc_library",
>     "drake_cc_googletest",
> ```

FYI -- alphasort.

---

*[drake/examples/kuka_iiwa_arm/pick_and_place/CMakeLists.txt, line 9 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/6199#-KlPykXyUHoGZBmhh30M:-KlPykXyUHoGZBmhh30N:b-hmbim5) ([raw file](https://github.com/robotlocomotion/drake/blob/05d27e3933ebc07da0bdeeec80b3d3f61ef67ee1/drake/examples/kuka_iiwa_arm/pick_and_place/CMakeLists.txt#L9)):*
> ```CMake
>       drakeLCMTypes
>       drakeLCMUtil
>       drakeRBM
> ```

FYI -- alphasort.

---

*[drake/examples/kuka_iiwa_arm/pick_and_place/world_state.cc, line 19 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/6199#-KlPzaz9seY5T6nnyqp1:-KlPzaz9seY5T6nnyqp2:b-j3mset) ([raw file](https://github.com/robotlocomotion/drake/blob/05d27e3933ebc07da0bdeeec80b3d3f61ef67ee1/drake/examples/kuka_iiwa_arm/pick_and_place/world_state.cc#L19)):*
> ```cpp
>   iiwa_end_effector_pose_ = Isometry3<double>::Identity();
>   iiwa_q_ = VectorX<double>::Zero(7);
>   iiwa_v_ = VectorX<double>::Zero(7);
> ```

BTW -- see note on the test cc; I think you should factor this 7.

---

*[drake/examples/kuka_iiwa_arm/pick_and_place/test/CMakeLists.txt, line 4 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/6199#-KlPzx7oXTesSnoIMKNx:-KlPzx7oXTesSnoIMKNy:b-hmbim5) ([raw file](https://github.com/robotlocomotion/drake/blob/05d27e3933ebc07da0bdeeec80b3d3f61ef67ee1/drake/examples/kuka_iiwa_arm/pick_and_place/test/CMakeLists.txt#L4)):*
> ```CMake
>   drake_add_cc_test(world_state_test)
>   target_link_libraries(world_state_test
>     pickAndPlaceCommon
> ```

FYI -- alphasort.

---

*[drake/examples/kuka_iiwa_arm/pick_and_place/test/world_state_test.cc, line 14 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/6199#-KlQ-0-lJnrAnHO4FtOm:-KlQ-0-lJnrAnHO4FtOn:btwc7ct) ([raw file](https://github.com/robotlocomotion/drake/blob/05d27e3933ebc07da0bdeeec80b3d3f61ef67ee1/drake/examples/kuka_iiwa_arm/pick_and_place/test/world_state_test.cc#L14)):*
> ```cpp
> namespace {
> 
> static const int kNumJoints = 7;
> ```

Given that this 7 is already in the cc,  I suggest factoring it to the `.h` file.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6199)*
<!-- Sent from Reviewable.io -->

sammy-tri commented:
Review status: 10 of 16 files reviewed at latest revision, 6 unresolved discussions.

---

*[drake/examples/kuka_iiwa_arm/dev/pick_and_place/pick_and_place_state_machine.cc, line 37 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/6199#-KlPwX7qu00RnKavMQHt:-KlQDUgrQMaOnbZKtzwM:bmq2ano) ([raw file](https://github.com/robotlocomotion/drake/blob/05d27e3933ebc07da0bdeeec80b3d3f61ef67ee1/drake/examples/kuka_iiwa_arm/dev/pick_and_place/pick_and_place_state_machine.cc#L37)):*
<details><summary><i>Previously, ggould-tri wrote…</i></summary><blockquote>

FYI -- This indentation is not GSG standard.
</blockquote></details>

Emacs is being weird about it .  Fixed.

---

*[drake/examples/kuka_iiwa_arm/pick_and_place/BUILD, line 8 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/6199#-KlPyWr0bw1MNRdfgm2x:-KlQConDx3vUwFi_c92C:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/05d27e3933ebc07da0bdeeec80b3d3f61ef67ee1/drake/examples/kuka_iiwa_arm/pick_and_place/BUILD#L8)):*
<details><summary><i>Previously, ggould-tri wrote…</i></summary><blockquote>

FYI -- alphasort.
</blockquote></details>

Done.

---

*[drake/examples/kuka_iiwa_arm/pick_and_place/CMakeLists.txt, line 9 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/6199#-KlPykXyUHoGZBmhh30M:-KlQCtCl29RXvwlLYtlg:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/05d27e3933ebc07da0bdeeec80b3d3f61ef67ee1/drake/examples/kuka_iiwa_arm/pick_and_place/CMakeLists.txt#L9)):*
<details><summary><i>Previously, ggould-tri wrote…</i></summary><blockquote>

FYI -- alphasort.
</blockquote></details>

Done.

---

*[drake/examples/kuka_iiwa_arm/pick_and_place/world_state.cc, line 19 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/6199#-KlPzaz9seY5T6nnyqp1:-KlQD7fd04guzhukSpRI:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/05d27e3933ebc07da0bdeeec80b3d3f61ef67ee1/drake/examples/kuka_iiwa_arm/pick_and_place/world_state.cc#L19)):*
<details><summary><i>Previously, ggould-tri wrote…</i></summary><blockquote>

BTW -- see note on the test cc; I think you should factor this 7.
</blockquote></details>

Done.

---

*[drake/examples/kuka_iiwa_arm/pick_and_place/test/CMakeLists.txt, line 4 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/6199#-KlPzx7oXTesSnoIMKNx:-KlQDDMLpqWzj-jyIIli:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/05d27e3933ebc07da0bdeeec80b3d3f61ef67ee1/drake/examples/kuka_iiwa_arm/pick_and_place/test/CMakeLists.txt#L4)):*
<details><summary><i>Previously, ggould-tri wrote…</i></summary><blockquote>

FYI -- alphasort.
</blockquote></details>

Done.

---

*[drake/examples/kuka_iiwa_arm/pick_and_place/test/world_state_test.cc, line 14 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/6199#-KlQ-0-lJnrAnHO4FtOm:-KlQDLiMlFUw4a7tEpp6:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/05d27e3933ebc07da0bdeeec80b3d3f61ef67ee1/drake/examples/kuka_iiwa_arm/pick_and_place/test/world_state_test.cc#L14)):*
<details><summary><i>Previously, ggould-tri wrote…</i></summary><blockquote>

Given that this 7 is already in the cc,  I suggest factoring it to the `.h` file.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6199)*
<!-- Sent from Reviewable.io -->

naveenoid commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: 10 of 16 files reviewed at latest revision, 2 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6199#-:-KlT8fw4Ou1dnnSzTHPM:bnfp4nl)*
<!-- Sent from Reviewable.io -->

ggould-tri commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 2 of 20 files at r1, 4 of 15 files at r2, 1 of 2 files at r4, 6 of 6 files at r5.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6199#-:-KlTlhG6OmORLb7IwOn3:bnfp4nl)*
<!-- Sent from Reviewable.io -->
